fdbclient/RESTSimKmsVault.actor.cpp

```diff
@@ -0,0 +1,510 @@
+
+		if (doc.HasMember(KMS_URLS_TAG) && doc[KMS_URLS_TAG].GetBool()) {
+			addKmsUrlsToDoc(result);
+		}
+
+		// Serialize json to string
+		rapidjson::StringBuffer sb;
+		rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+		doc.Accept(writer);
+
+		response->code = HTTP::HTTP_STATUS_CODE_OK;
+		response->data.headers = request->data.headers;
+		PacketWriter pw(response->data.content->getWriteBuffer(sb.GetSize()), nullptr, Unversioned());
+		pw.serializeBytes(sb.GetString(), sb.GetSize());
+		response->data.contentLen = sb.GetSize();
+	} catch (Error& e) {
+		TraceEvent("FetchByDomainIdsFailed").error(e);
+		throw;
+	}
+}
+
+void handleFetchKeysByKeyIds(Reference<HTTP::IncomingRequest> request, Reference<HTTP::OutgoingResponse> response) {
+	rapidjson::Document doc;
+	doc.Parse(request->data.content.data());
+
+	try {
+		const int version = extractVersion(request, response, doc);
+
+		checkValidationTokens(request, response, version, doc);
+
+		rapidjson::Document result;
+		result.SetObject();
+
+		// Append 'request version'
+		addVersionToDoc(result, version);
+
+		// Append 'cipher_key_details' as json array
+		rapidjson::Value cipherDetails(rapidjson::kArrayType);
+		for (const auto& cipherDetail : doc[CIPHER_KEY_DETAILS_TAG].GetArray()) {
+			Optional<EncryptCipherDomainId> domainId;
+			if (cipherDetail.HasMember(ENCRYPT_DOMAIN_ID_TAG) && cipherDetail[ENCRYPT_DOMAIN_ID_TAG].IsInt64()) {
+				domainId = cipherDetail[ENCRYPT_DOMAIN_ID_TAG].GetInt64();
+			}
+			EncryptCipherBaseKeyId baseCipherId = cipherDetail[BASE_CIPHER_ID_TAG].GetUint64();
+			Reference<SimKmsVaultKeyCtx> keyCtx = SimKmsVault::getByBaseCipherId(baseCipherId);
+			addCipherDetailToRespDoc(doc, cipherDetails, keyCtx, domainId);
+		}
+		rapidjson::Value memberKey(CIPHER_KEY_DETAILS_TAG, doc.GetAllocator());
+		result.AddMember(memberKey, cipherDetails, doc.GetAllocator());
+
+		if (doc.HasMember(KMS_URLS_TAG) && doc[KMS_URLS_TAG].GetBool()) {
+			addKmsUrlsToDoc(result);
+		}
+
+		// Serialize json to string
+		rapidjson::StringBuffer sb;
+		rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+		doc.Accept(writer);
+
+		response->code = HTTP::HTTP_STATUS_CODE_OK;
+		response->data.headers = request->data.headers;
+		PacketWriter pw(response->data.content->getWriteBuffer(sb.GetSize()), nullptr, Unversioned());
+		pw.serializeBytes(sb.GetString(), sb.GetSize());
+		response->data.contentLen = sb.GetSize();
+	} catch (Error& e) {
+		TraceEvent("FetchByKeyIdsFailed").error(e);
+		throw;
+	}
+}
+
+ACTOR Future<Void> simKmsVaultRequestHandler(Reference<HTTP::IncomingRequest> request,
+                                             Reference<HTTP::OutgoingResponse> response) {
+	wait(delay(0));
+	ASSERT_EQ(request->verb, HTTP::HTTP_VERB_POST);
+
+	if (request->resource.compare("/get-encryption-keys-by-key-ids") == 0) {
+		handleFetchKeysByKeyIds(request, response);
+	} else if (request->resource.compare("/get-encryption-keys-by-domain-ids") == 0) {
```

Comment:
Are we intending to handle blob metadata in this PR or a separate one?
Switching simulation to use this by default would break blob tests without that implemented, for example

Reply:
Ok, I see